[Sean Leong] iP

seanleong339 opened the pull request:
# Duke
> One of the most dangerous forms of human error is forgetting what one is trying to achieve. - Paul Nitze [Source](https://www.brainyquote.com/quotes/paul_nitze_205927?src=t_forgetting)

Never forget anything, or at least anything you write in Duke! It's,
- In English
- Simple to learn
- ~FAST~ *VERY* FAST to use

All you have to do is,

1. Download it from [here](https://github.com/seanleong339/ip/releases/tag/A-Jar)
2. Add your tasks
3. Let it manage your tasks for you 🤪 

And it is **FREE OF CHARGE**!

Features:
- [x] 3 kinds of tasks
- [x] Reminder of tasks due daily
- [x] Date due!

Some basic commands are:
```
todo
deadline
event
today
bye
```

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

seanleong339 committed:
level 1

seanleong339 committed:
level 2

seanleong339 committed:
level 3

seanleong339 committed:
level 4, added DukeList class to manage ds methods, and subclass of Tasks

seanleong339 committed:
level 5, added custom exception

seanleong339 committed:
level-6. Added method to delete, documentation of methods in DukeList

seanleong339 committed:
Change packages to import

seanleong339 committed:
add Level 7 changes

seanleong339 committed:
add Level 8 changes and stretch goal

seanleong339 committed:
add msg for no task due

seanleong339 committed:
Merge branch 'branch-Level-7'

seanleong339 committed:
remove class files

seanleong339 committed:
Resolve merge conflicts

seanleong339 committed:
fix missing exception

seanleong339 committed:
Remove ignored files

seanleong339 committed:
Remove starting logo

seanleong339 committed:
Refactor code to use more OOP and stretch goal

seanleong339 committed:
Refactor code into packages

seanleong339 committed:
Add JUnit tests

seanleong339 committed:
Change file path for storage

seanleong339 committed:
Update gitignore

seanleong339 committed:
Update gitignore

seanleong339 committed:
Add comments to methods

seanleong339 committed:
Tweak code to apply with 2103 coding standard

seanleong339 committed:
Add find method

seanleong339 committed:
Merge branch 'branch-A-JavaDoc'

seanleong339 committed:
Fix merge conflicts

seanleong339 committed:
Fix merge conflicts

seanleong339 committed:
Remove ignore files

AAlghrairy commented on src/main/java/duke/commands/AddDeadline.java:
```diff
@@ -0,0 +1,41 @@
+    public void execute(Ui ui, DukeList list) {
+        Deadlines ans = new Deadlines(msg, date);
+        list.add(ans);
+    }
```
Any reason you did not delegate the process of adding tasks to the list to your DukeList class? Also, any reason you did not have just one "AddTask" command class, rather than one for each task type?

I've noticed this issue in several other places.

AAlghrairy commented on src/main/java/duke/commands/TryAgain.java:
```diff
@@ -0,0 +1,20 @@
+import duke.tasklist.DukeList;
+import duke.ui.Ui;
+
+public class TryAgain extends Command {
```
Perhaps add a header for this class (as it is a bit unclear what this class is for)?

AAlghrairy commented on src/main/java/duke/commands/TryAgain.java:
```diff
@@ -0,0 +1,20 @@
+public class TryAgain extends Command {
+
+    @Override
+    public void execute(Ui ui, DukeList list) {
+    }
```
Any reason why this method body is empty? Consider adding a header comment to explain why.

seanleong339 replied:
It was cause I needed to return a Command object regardless from the method in my Parser class, so this command is used when an invalid command is input by the user. Its a design flaw on my part, will think of ways to fix it.

AAlghrairy commented on src/main/java/duke/parser/DukeParser.java:
```diff
@@ -0,0 +1,202 @@
+            }
+        }
+        return ans;
+    }
```
Perhaps delegate this process to the Storage class instead (since its purpose is to deal with File processes)?

AAlghrairy commented on src/main/java/duke/parser/DukeParser.java:
```diff
@@ -0,0 +1,202 @@
+            } else {
+                return new AddToDos(arg[1]);
+            }
+        } else if (cmd.equals("deadline")) {
+            if (arg.length == 1) {
+                throw new MissingArgumentException();
+            } else if (!arg[1].matches(".+/by.+")) {
+                throw new WrongFormatException();
+            } else {
+                String[] body = arg[1].split("/by");
+                if (!isValidDate(body[1].trim())) {
+                    throw new InvalidDateException();
+                } else {
+                    return new AddDeadline(body[0], body[1].trim());
+                }
+            }
+        } else if (cmd.equals("event")) {
+            if (arg.length == 1) {
+                throw new MissingArgumentException();
+            } else if (!arg[1].matches(".+/at.+")) {
+                throw new WrongFormatException();
+            } else {
+                String[] body = arg[1].split("/at");
+                if (!isValidDate(body[1].trim())) {
+                    throw new InvalidDateException();
+                } else {
+                    return new AddEvents(body[1], body[1].trim());
+                }
+            }
+        } else if (cmd.equals("find")) {
+            if (arg.length == 1) {
+                throw new MissingArgumentException();
+            } else {
+                return new FindCommand(arg[1]);
+            }
+        } else {
+            throw new InvalidCommandException();
+        }
+
+    }
```
Any reason you did not use Enums to specify the type of each command (e.g. Creating an Enum class for task types) rather than repeatedly using .equals() command?

Also, any reason you did not use a switch statement here instead of repeatedly using "if" and "else if"?

seanleong339 replied:
Was not familiar with Enums previously, will be adding your suggestions in subsequent commits.

AAlghrairy commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+
+    private String s;
+    private boolean isDone;
+    private LocalDate date;
```
`private LocalDate date;`

Consider deleting this variable (as well as its respective getter and setter methods) since it is only utilized in Events and Deadline child classes (It is good practice to make sure that variables and methods in a parent class are only included if they will be shared across all the parent class' respective child classes).

AAlghrairy commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,25 @@
+import duke.ui.Ui;
+
+
+public class Duke {
```
Perhaps add a header comment for each of the classes?

I noticed this issue in the other class files too.

AAlghrairy commented on src/main/java/duke/tasklist/DukeList.java:
```diff
@@ -0,0 +1,131 @@
+    /**
+     * Prints a list of tasks due on current date to the console
+     */
+    public void todayTask(){
```
```suggestion
    public void getTodayTask() {
```

AAlghrairy commented on src/main/java/duke/tasklist/DukeList.java:
```diff
@@ -0,0 +1,131 @@
+        LocalDate cur = LocalDate.now();
+        String day = "\nDuke: Here are the tasks due today\n";
+        boolean b = true;
+        for(Task t: a){
```
```suggestion
        for(Task t: a) {
```

AAlghrairy commented on src/main/java/duke/test.java:
```diff
@@ -0,0 +1,16 @@
+import java.io.IOException;
+import java.util.Scanner;
+
+public class test {
```
```suggestion
public class Test {
```

AAlghrairy commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+        } else {
+            return "[ ] " + this.s;
+        }
+    }
```
Perhaps override the "toString" method for processes where you need to show the instance variables? 
```suggestion
    @Override
    public String toString() {
        if(isDone) {
            return "[X] " + this.s;
        } else {
            return "[ ] " + this.s;
        }
    }
```
I noticed this in the other Task child classes too.

AAlghrairy reviewed (commented):
Overall, I found the code to be alright, although  I noted some OOP violations in certain areas (e.g. Parser class performing the action of reading Files, which should be a function in the Storage class). I also noted that there were no header comments in the Classes. Consider using Enums for the task types in lieu of using .equals(), if that makes sense.

AAlghrairy commented on src/main/java/duke/tasklist/DukeList.java:
```diff
@@ -0,0 +1,131 @@
+
+public class DukeList {
+
+    private List<Task> a;
```
Perhaps rename this variable to be more descriptive about it's usage (e.g. "tasks)?

AAlghrairy commented on src/main/java/duke/tasklist/DukeList.java:
```diff
@@ -0,0 +1,131 @@
+    public void add(Task t) {
+        a.add(t);
+        storage.store(a);
+        System.out.println("\nDuke: Got it. I've added this task:\n      "
+                + t.show() + "\n      Now you have "
+                + this.a.size() + " tasks in the list.\n");
```
Perhaps delegate the process of outputting messages to the UI class (as this could violate OOP principle)? I've noticed this issue in the other methods of this class as well.

seanleong339 replied:
Hi, really appreciate all the comments that you've left, I think they were really helpful for me to identify the general problems with my code design and how I should improve. Will probably take me some time to fix the code design issues but thanks a lot!

AAlghrairy replied:
No problem. Glad it could help 👍

AAlghrairy reviewed (commented):
_Additional feedback_

seanleong339 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2122S2/ip

seanleong339 committed:
Set up Gradle

seanleong339 committed:
Merge branch 'A-Gradle'

seanleong339 committed:
Fix code to match coding standards

seanleong339 committed:
Add GUI to Duke

seanleong339 committed:
Merge branch 'branch-Level-10'

seanleong339 committed:
Edit MainWindow.fxml

seanleong339 committed:
Add assertions in the code

seanleong339 committed:
Refactor code to improve the code quality

createCommand method in Parser is too long, and contains many repeated
lines of code. This is the same for the todayTask method in
DukeList.

Extracting the code logic allows us to use abstraction to  reduce
the amount of repeated code, and also increases the code readability.

Let's write new methods that contain the same code logic for different
commands in DukeParser.

Abstraction is preferable as code logic for certain commands with same
checks can be easily changed.

seanleong339 committed:
Merge pull request #1 from seanleong339/branch-A-Assertions

Add assertions in the code

seanleong339 committed:
Merge branch 'master' into branch-A-CodeQuality

seanleong339 committed:
Merge pull request #2 from seanleong339/branch-A-CodeQuality

Refactor code to improve the code quality

seanleong339 committed:
Remove unused variables

seanleong339 committed:
Remove redundant readInput method from UI

seanleong339 committed:
Give credit in comment

seanleong339 committed:
Add feature that prevents adding duplicate task

seanleong339 committed:
Merge branch 'branch-C-DetectDuplicates'

seanleong339 committed:
Change icons and gui title

seanleong339 committed:
Add User guide

seanleong339 committed:
Change GUI

seanleong339 committed:
Rename product snapshot

seanleong339 committed:
Rename Ui snapshot

seanleong339 committed:
Remove wrong syntax in MainWindow.fxml
